qiskit/aqua/operators/gradients/circuit_gradients/lin_comb.py

```diff
@@ -0,0 +1,791 @@
+                            else:
+                                self.insert_gate(grad_state, param_occurence[0],
+                                                 gate_to_insert,
+                                                 additional_qubits=additional_qubits)
+                        else:
+                            self.insert_gate(grad_state, param_occurence[0],
+                                             gate_to_insert,
+                                             additional_qubits=additional_qubits)
+                        grad_state.h(work_q)
+
+                        state = np.sqrt(np.abs(coeff_i)) * state_op.coeff * CircuitStateFn(
+                            grad_state)
+                        # Chain Rule parameter expressions
+                        gate_param = param_occurence[0].params[param_occurence[1]]
+                        if meas_op:
+                            if gate_param == param:
+                                state = meas_op @ state
+                            else:
+                                if isinstance(gate_param, ParameterExpression):
+                                    expr_grad = DerivativeBase.parameter_expression_grad(gate_param,
+                                                                                         param)
+                                    state = (expr_grad * meas_op) @ state
+                                else:
+                                    state = ~Zero @ One
+                        else:
+                            if gate_param == param:
+                                state = ListOp([state],
+                                               combo_fn=partial(self._grad_combo_fn,
+                                                                state_op=state_op))
+                            else:
+                                if isinstance(gate_param, ParameterExpression):
+                                    expr_grad = DerivativeBase.parameter_expression_grad(gate_param,
+                                                                                         param)
+                                    state = expr_grad * ListOp(
+                                        [state],
+                                        combo_fn=partial(self._grad_combo_fn, state_op=state_op))
+                                else:
+                                    state = ~Zero @ One
+
+                        if m == 0 and k == 0:
+                            op = state
+                        else:
+                            # Product Rule
+                            op += state
+                if len(target_params) > 1:
+                    if not states:
+                        states = [op]
+                    else:
+                        states += [op]
+                else:
+                    return op
+        if len(target_params) > 1:
+            return ListOp(states)
+        else:
+            return op
```

Comment:
So the gradient computations tend to take quite a while -- have we tried calling `reduce` on the operators before returning them? Not sure though if the overhead of reducing them is larger than the additional time in evaluating them.

Reply:
We do call `reduce` in the parameter shift case. 
Haven't tried it for the linear combination case. Not sure if `reduce` preserves the form that is needed for this implementation.

Reply:
Ok, if it might break things nevermind 😄 then we can do it later.